cloudinit/sources/helpers/vultr.py

```diff
@@ -0,0 +1,293 @@
+
+import json
+import copy
+import base64
+
+from cloudinit import log as log
+from cloudinit import url_helper
+from cloudinit import dmi
+from cloudinit import util
+from cloudinit import net
+from cloudinit.net.dhcp import EphemeralDHCPv4, NoDHCPLeaseError
+from functools import lru_cache
+
+# Get LOG
+LOG = log.getLogger(__name__)
+
+
+@lru_cache()
+def get_metadata(params):
+    params = json.loads(params)
+
+    # Bring up interface
+    try:
+        with EphemeralDHCPv4(connectivity_url=params['url']):
+            # Fetch the metadata
+            v1 = fetch_metadata(params)
+    except (NoDHCPLeaseError) as exc:
+        LOG.error("Bailing, DHCP Exception: %s", exc)
+        raise
+
+    v1_json = json.loads(v1)
+    metadata = v1_json
+
+    # This comes through as a string but is JSON, make a dict
+    metadata['vendor-config'] = json.loads(metadata['vendor-config'])
+
+    return json.dumps(metadata)
+
+
+def get_cached_metadata(args):
+    return json.loads(get_metadata(json.dumps(args)))
+
+
+# Read the system information from SMBIOS
+def get_sysinfo():
+    return {
+        'manufacturer': dmi.read_dmi_data("system-manufacturer"),
+        'subid': dmi.read_dmi_data("system-serial-number")
+    }
+
+
+# Assumes is Vultr is already checked
+def is_baremetal():
+    if get_sysinfo()['manufacturer'] != "Vultr":
+        return True
+    return False
+
+
+# Confirm is Vultr
+def is_vultr():
+    # VC2, VDC, and HFC use DMI
+    sysinfo = get_sysinfo()
+
+    if sysinfo['manufacturer'] == "Vultr":
+        return True
+
+    # Baremetal requires a kernel parameter
+    if "vultr" in util.get_cmdline():
+        return True
+
+    return False
+
+
+def convert_to_base64(string):
+    string_bytes = string.encode('ascii')
+    b64_bytes = base64.b64encode(string_bytes)
+    return b64_bytes.decode('ascii')
+
+
+# Read Metadata endpoint
+def read_metadata(params):
+    response = url_helper.readurl(params['url'],
+                                  timeout=params['timeout'],
+                                  retries=params['retries'],
+                                  headers={'Metadata-Token': 'vultr'},
+                                  sec_between=params['wait'])
+
+    if not response.ok():
+        raise RuntimeError("Failed to connect to %s: Code: %s" %
+                           params['url'], response.code)
+
+    return response.contents.decode()
+
+
+# Get Metadata by flag
+def fetch_metadata(params):
+    req = dict(params)
+    req['url'] = "%s/v1.json" % params['url']
+
+    return read_metadata(req)
+
+
+# Wrapped for caching
+@lru_cache()
+def get_interface_map():
+    return net.get_interfaces_by_mac()
+
+
+# Convert macs to nics
+def get_interface_name(mac):
+    macs_to_nic = get_interface_map()
+
+    if mac not in macs_to_nic:
+        return None
+
+    return macs_to_nic.get(mac)
+
+
+# Generate network configs
+def generate_network_config(config):
```

Comment:
the only thing this uses 'config' for is to call get_cached_metadata().  How about just having the caller pass in the 'md' variable ?